application/src/main/java/org/thingsboard/server/service/system/DefaultSystemInfoService.java

```diff
@@ -0,0 +1,207 @@
+
+@TbCoreComponent
+@Service
+@RequiredArgsConstructor
+@Slf4j
+public class DefaultSystemInfoService extends TbApplicationEventListener<PartitionChangeEvent> implements SystemInfoService {
+
+    public static final FutureCallback<Integer> CALLBACK = new FutureCallback<>() {
+        @Override
+        public void onSuccess(@Nullable Integer result) {
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            log.warn("Failed to persist system info", t);
+        }
+    };
+
+    private final TbServiceInfoProvider serviceInfoProvider;
+    private final PartitionService partitionService;
+    private final DiscoveryService discoveryService;
+    private final TelemetrySubscriptionService telemetryService;
+    private final TbApiUsageStateClient apiUsageStateClient;
+    private ScheduledExecutorService scheduler;
+
+    @Value("${zk.enabled:false}")
+    private boolean zkEnabled;
+
+    @Override
+    protected void onTbApplicationEvent(PartitionChangeEvent partitionChangeEvent) {
+        if (ServiceType.TB_CORE.equals(partitionChangeEvent.getServiceType())) {
+            if (scheduler == null && partitionService.resolve(ServiceType.TB_CORE, TenantId.SYS_TENANT_ID, TenantId.SYS_TENANT_ID).isMyPartition()) {
+                scheduler = Executors.newSingleThreadScheduledExecutor(ThingsBoardThreadFactory.forName("tb-system-info-scheduler"));
+                scheduler.scheduleAtFixedRate(this::saveCurrentSystemInfo, 0, 1, TimeUnit.MINUTES);
+            } else {
+                destroy();
+            }
+        }
+    }
+
+    @Override
+    public SystemInfo getSystemInfo() {
+        SystemInfo systemInfo = new SystemInfo();
+
+        ServiceInfo serviceInfo = serviceInfoProvider.getServiceInfoWithCurrentSystemInfo();
+
+        if (zkEnabled) {
+            systemInfo.setSystemData(getSystemData(serviceInfo));
+        } else {
+            systemInfo.setMonolith(true);
+            systemInfo.setSystemData(Collections.singletonList(createSystemInfoData(serviceInfo)));
+        }
+
+        return systemInfo;
+    }
+
+    protected void saveCurrentSystemInfo() {
+        if (zkEnabled) {
+            saveCurrentClusterSystemInfo();
+        } else {
+            saveCurrentMonolithSystemInfo();
+        }
+    }
+
+    private void saveCurrentClusterSystemInfo() {
+        long ts = System.currentTimeMillis();
+
+        List<SystemInfoData> clusterSystemData = getSystemData(serviceInfoProvider.getServiceInfoWithCurrentSystemInfo());
+        BasicTsKvEntry clusterDataKv = new BasicTsKvEntry(ts, new JsonDataEntry("clusterSystemData", JacksonUtil.toString(clusterSystemData)));
+        doSave(Collections.singletonList(clusterDataKv));
+    }
+
+    private void saveCurrentMonolithSystemInfo() {
+        long ts = System.currentTimeMillis();
+        List<TsKvEntry> tsList = new ArrayList<>();
+
+        Long memoryUsage = getMemoryUsage();
```

Comment:
Code will be much cleaner if getMemoryUsage returns Optional. Smth like `getMemoryUsage().map().ifPresent(tsList::add)`